llvm/lib/Transforms/InstCombine/InstCombineSelect.cpp

```diff
@@ -2793,7 +2793,11 @@ static Instruction *foldSelectWithFCmpToFabs(SelectInst &SI,
 
     // fold (X <= +/-0.0) ? (0.0 - X) : X to fabs(X), when 'Swap' is false
     // fold (X >  +/-0.0) ? X : (0.0 - X) to fabs(X), when 'Swap' is true
-    if (match(TrueVal, m_FSub(m_PosZeroFP(), m_Specific(X)))) {
+    // Note: We require "nnan" for this fold because fcmp ignores the signbit
+    //       of NAN, but IEEE-754 specifies the signbit of NAN values with
+    //       fneg/fabs operations.
+    if (cast<FPMathOperator>(CondVal)->hasNoNaNs() &&
+        match(TrueVal, m_FSub(m_PosZeroFP(), m_Specific(X)))) {
```

Comment:
Can't you equivalently check the nnan on the fsub? or check isKnownNeverNan on X?

Reply:
> or check isKnownNeverNan on X?

Done.

Reply:
nnan on the select also works? https://alive2.llvm.org/ce/z/R57ux5

Reply:
Done